File: 2015/go/07.go
// Advent of Code 2015, day 7
// (c) aichingert

package main

import (
	"fmt"
	"os"
	"strings"
	"strconv"
)

func solve(cond map[string]string, vals map[string]int, cur string) int {
	res, err := strconv.Atoi(cur)
	
	if err == nil {
		return res
	}

	val, ok := vals[cur]

	if ok {
		return val
	}

	s := strings.Split(cond[cur], " ")
	ret := 0

	switch len(s) {
	case 1:
		ret = solve(cond, vals, s[0])
	case 2:
		ret = ^solve(cond, vals, s[1])
	case 3:
		switch s[1] {
		case "AND":
			ret = solve(cond, vals, s[0]) & solve(cond, vals, s[2])
		case "OR":
			ret = solve(cond, vals, s[0]) | solve(cond, vals, s[2])
		case "LSHIFT":
			shift,_ := strconv.Atoi(s[2])
			ret = solve(cond, vals, s[0]) << shift
		case "RSHIFT":
			shift,_ := strconv.Atoi(s[2])
			ret = solve(cond, vals, s[0]) >> shift
		}
	}

	vals[cur] = ret
	return ret
}

func parse(lines []string) map[string]string {
	vals := make(map[string]string)

	for i := range lines {
		splt := strings.Split(lines[i], " -> ")
		vals[splt[1]] = splt[0]
	}

	return vals
}

func main() {
	dat, err := os.ReadFile("../input/07")

	if err != nil {
		panic(err)
	}

	lines := strings.Split(strings.ReplaceAll(string(dat), "\r\n", "\n"), "\n")
	cond := parse(lines)
	vals := make(map[string]int)

	a := solve(cond, vals, "a")
	for k := range vals { delete(vals,k) }
	vals["b"] = a

	fmt.Printf("Part 1: %d\n", a)
	fmt.Printf("Part 2: %d\n", solve(cond, vals, "a"))
}

File: 2015/julia/07.jl
# Advent of Code 2015, day 7
# (c) aichingert

module Day7

function solve(instr::Dict{Any,Any}, cur::String)
    if tryparse(UInt16, cur) !== nothing
	return parse(UInt16, cur)
    end

    if cur in keys(calc) return calc[cur] end

    values = split(instr[cur], " ")

    if length(values) == 1 global calc[cur] = solve(instr,String(values[1]))
    elseif length(values) == 2 global calc[cur] = ~solve(instr,String(values[2])) 
    else
        if values[2] == "AND" global calc[cur] = solve(instr,String(values[1])) & solve(instr,String(values[3]))
	elseif values[2] == "OR" global calc[cur] = solve(instr,String(values[1])) | solve(instr,String(values[3]))
	elseif values[2] == "LSHIFT" global calc[cur] = solve(instr,String(values[1])) << solve(instr,String(values[3]))
    	else global calc[cur] = solve(instr,String(values[1])) >> solve(instr,String(values[3])) end
    end

    return calc[cur] 
end

instr = Dict()
calc = Dict()

for l in eachline("../input/07")
    s = split(l, " -> ")
    instr[s[end]] = s[begin]
end

part1 = solve(instr, "a")
empty!(calc)
calc["b"] = part1

println("Part 1: " * string(part1))
println("Part 2: " * string(solve(instr, "a")))

end
